### PR TITLE
Backup: add checkpoint

### LINKED ISSUE #38647: Backup: add checkpoint for backup
## Enhancement
Some errors in the backup process will cause the backup to be interrupted. If the errors can be recovered, such as one TiKV's hard disk is full, we can try the next backup after the errors are corrected. But for a large-scale (200 TB) cluster, it costs too much. Therefore, the checkpoint feature is required.

### PR DESCRIPTION
**Pull request by Leavrth**
Signed-off-by: Leavrth <jianjun.liao@outlook.com>

### What problem does this PR solve?

Issue Number: close #38647 

Problem Summary: add checkpoint for backup

### What is changed and how it works?
Record the backed up region information at regular intervals.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/task/backup.go [hunk 1]
```diff
@@ -296,7 +328,11 @@ func RunBackup(c context.Context, g glue.Glue, cmdName string, cfg *BackupConfig
 	// Domain loads all table info into memory. By skipping Domain, we save
 	// lots of memory (about 500MB for 40K 40 fields YCSB tables).
 	needDomain := !skipStats
-	mgr, err := NewMgr(ctx, g, cfg.PD, cfg.TLS, GetKeepalive(&cfg.Config), cfg.CheckRequirements, needDomain, conn.NormalVersionChecker)
+	versionCheckerNum := conn.NormalVersionChecker
+	if cfg.UseCheckpoint {
```
Comment (3pointer) [inline br/pkg/task/backup.go:332]: it seems we lost some check when using checkpoint?

File: br/pkg/task/backup.go [hunk 2]
```diff
@@ -324,6 +360,10 @@ func RunBackup(c context.Context, g glue.Glue, cmdName string, cfg *BackupConfig
 	if err != nil {
 		return errors.Trace(err)
 	}
+
+	// set cipher only for checkpoint
+	client.SetCipher(&cfg.CipherInfo)
```
Comment (3pointer) [inline br/pkg/task/backup.go:365]: why need this?

File: br/pkg/task/backup.go [hunk 3]
```diff
@@ -150,10 +156,19 @@ func (cfg *BackupConfig) ParseFromFlags(flags *pflag.FlagSet) error {
...
+		return errors.Trace(err)
+	}
 	gcTTL, err := flags.GetInt64(flagGCTTL)
 	if err != nil {
 		return errors.Trace(err)
 	}
+	// if use checkpoint and gcTTL is the default value
+	// update gcttl to 1 hours
+	if cfg.UseCheckpoint && gcTTL == utils.DefaultBRGCSafePointTTL {
+		gcTTL = 3600
```
Comment (3pointer) [inline br/pkg/task/backup.go:170]: maybe add some log here if we change the default configuration.

File: br/pkg/checkpoint/checkpoint.go [hunk 4]
```diff
@@ -0,0 +1,528 @@
+package checkpoint
```
Comment (3pointer) [inline br/pkg/checkpoint/checkpoint.go:1]: miss license

File: br/pkg/checkpoint/checkpoint_test.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
+package checkpoint_test
```
Comment (3pointer) [inline br/pkg/checkpoint/checkpoint_test.go:1]: ditto

File: br/pkg/task/backup.go (hunk 1 above)
Comment (Leavrth) [reply br/pkg/task/backup.go]: done. change to mark the support error in version pkg.

File: br/pkg/task/backup.go (hunk 2 above)
Comment (Leavrth) [reply br/pkg/task/backup.go:371]: encrypt the key info in the checkpoint files if in need

File: br/pkg/task/backup.go (hunk 3 above)
Comment (Leavrth) [reply br/pkg/task/backup.go]: done.

File: br/pkg/checkpoint/checkpoint.go (hunk 4 above)
Comment (Leavrth) [reply br/pkg/checkpoint/checkpoint.go:15]: done.

File: br/pkg/checkpoint/checkpoint_test.go (hunk 5 above)
Comment (Leavrth) [reply br/pkg/checkpoint/checkpoint_test.go:15]: done.

File: br/pkg/task/backup.go [hunk 6]
```diff
@@ -353,7 +412,26 @@ func RunBackup(c context.Context, g glue.Glue, cmdName string, cfg *BackupConfig
...
 	log.Info("current backup safePoint job", zap.Object("safePoint", sp))
-	err = utils.StartServiceSafePointKeeper(ctx, mgr.GetPDClient(), sp)
+	cctx, gcSafePointKeeperCancel := context.WithCancel(ctx)
+	gcSafePointKeeperRemovable := false
+	defer func() {
+		// don't reset the gc-safe-point if checkpoint mode is used and backup is not finished
+		if cfg.UseCheckpoint && !gcSafePointKeeperRemovable {
+			return
+		}
+		log.Info("start to remove gc-safepoint.")
```
Comment (3pointer) [inline br/pkg/task/backup.go:422]:
```suggestion
		log.Info("start to remove gc-safepoint keeper.")
```

File: br/pkg/task/backup.go [hunk 7]
```diff
@@ -353,7 +412,26 @@ func RunBackup(c context.Context, g glue.Glue, cmdName string, cfg *BackupConfig
...
+		// close the gc safe point keeper at first
+		gcSafePointKeeperCancel()
+		// set the ttl to 0 to remove the gc-safe-point
+		sp.TTL = 0
+		if err := utils.UpdateServiceSafePoint(ctx, mgr.GetPDClient(), sp); err != nil {
+			log.Warn("failed to update service safe point, backup may fail if gc triggered",
+				zap.Error(err),
+			)
+		}
+		log.Info("finish removing gc-safepoint.")
```
Comment (3pointer) [inline br/pkg/task/backup.go:432]:
```suggestion
		log.Info("finish removing gc-safepoint keeper.")
```

File: br/pkg/task/backup.go [hunk 8]
```diff
@@ -301,6 +338,13 @@ func RunBackup(c context.Context, g glue.Glue, cmdName string, cfg *BackupConfig
 		return errors.Trace(err)
 	}
 	defer mgr.Close()
+	// after version check, check the cluster whether support checkpoint mode
+	if cfg.UseCheckpoint {
```
Comment (3pointer) [inline br/pkg/task/backup.go:342]: shall we fall back to normal backup without checkpoint?

Comment (3pointer) [reply br/pkg/task/backup.go:342]: And it looks like the default value of UseCheckpoint is true. then it must failed when run backup on v6.4.0?

File: br/pkg/task/backup.go [hunk 9]
```diff
@@ -45,11 +48,13 @@ const (
 	flagRemoveSchedulers = "remove-schedulers"
 	flagIgnoreStats      = "ignore-stats"
 	flagUseBackupMetaV2  = "use-backupmeta-v2"
+	flagNoCheckpoint     = "no-checkpoint"
```
Comment (3pointer) [inline br/pkg/task/backup.go:51]: maybe rename it to flagUseCheckpoint and set default value to true.

Comment (3pointer) [reply br/pkg/task/backup.go:51]: don't let user use double negative grammer

File: br/pkg/backup/schema.go [hunk 10]
```diff
@@ -100,6 +110,11 @@ func (ss *Schemas) BackupSchemas(
 			schema.dbInfo.Name = utils.TemporaryDBName(schema.dbInfo.Name.O)
 		}
 
+		var checksum *checkpoint.ChecksumItem
```
Comment (3pointer) [inline br/pkg/backup/schema.go:113]: I think we should move this into `WorkerPool`

File: br/pkg/task/backup.go (hunk 9 above)
Comment (Leavrth) [reply br/pkg/task/backup.go]: done

File: br/pkg/task/backup.go (hunk 8 above)
Comment (Leavrth) [reply br/pkg/task/backup.go:342]: done, print the error to log, and fall back to normal backup mode

File: br/pkg/backup/schema.go (hunk 10 above)
Comment (Leavrth) [reply br/pkg/backup/schema.go:113]: we define it here to avoid the data race of map `Schemas::checkpointChecksum`.

**Review by 3pointer — APPROVED** [review]

**Review by YuJuncen — APPROVED** [review]
rest lgtm

**Review by joccau — APPROVED** [review]
LGTM

File: br/pkg/checkpoint/checkpoint.go [hunk 11]
```diff
@@ -0,0 +1,580 @@
...
+					if err := r.flushMeta(ctx, errCh); err != nil {
+						r.errCh <- err
+					}
+					// close the channel to flush worker
+					// and wait it to consumes all the metas
+					close(r.metaCh)
+					wg.Wait()
+					return
+				}
+				groups, exist := r.meta[msg.GroupKey]
```
[the remark this answers is not in the document]
Comment (Leavrth) [reply br/pkg/checkpoint/checkpoint.go:377]: `flushWorker` is only called in the `checkpointLoop`. And there is only this goruntine access it.

File: br/pkg/checkpoint/checkpoint.go [hunk 12]
```diff
@@ -0,0 +1,580 @@
...
+// start a goroutine to flush the meta, which is sent from `checkpoint looper`, to the external storage
+func (r *CheckpointRunner) startCheckpointRunner(ctx context.Context, wg *sync.WaitGroup) chan error {
+	errCh := make(chan error)
+	wg.Add(1)
+	flushWorker := func(ctx context.Context, errCh chan error) {
+		defer wg.Done()
+		for {
+			select {
+			case <-ctx.Done():
+
```
[the remark this answers is not in the document]
Comment (Leavrth) [reply br/pkg/checkpoint/checkpoint.go]: done

File: br/pkg/checkpoint/checkpoint.go [hunk 13]
```diff
@@ -0,0 +1,580 @@
...
+	var fname []byte = nil
+
+	for _, group := range meta {
+		if len(group.Groups) == 0 {
+			continue
+		}
+
+		// use the first item's group-key and sub-range-key as the filename
+		if len(fname) == 0 {
+			fname = append([]byte(group.GroupKey+".."), group.Groups[0].StartKey...)
```
[the remark this answers is not in the document]
Comment (Leavrth) [reply br/pkg/checkpoint/checkpoint.go]: done

**Pull request merged**